### ISSUE TITLE
Relationship between Lexicon and LM vocab

### ISSUE THREAD

**Issue by abhinavkulkarni** [facebookresearch/wav2letter#913]
Hi,

For streaming inference examples, what is the relationship between lexicon and KenLM vocab? The [downloaded files from the wiki](https://github.com/facebookresearch/wav2letter/wiki/Inference-Run-Examples#download-the-example-trained-models-from-aws-s3) have a token, lexicon, and KenLM file. The `LexiconDecoder` can only decode words present in the Lexicon (200k in number). In it, the partial beams are scored based on the prefixes of incomplete words (by using max smearing). Whenever a full word is found, the max-smeared score is removed and the word score from LM is added.

Given this, does the LM need to store any other words that are not in Lexicon? 

I am imagining, the LM could be trained on a bigger corpus to correctly capture n-gram probabilities and then only those words from Lexicon were retained to reduce LM size.

I haven't read the KenLM paper but do backoff and smoothing require retaining other non-Lexicon words in LM?

Thanks!

**Comment by tlikhomanenko** [comment]
For smearing we are using only unigrams scores. Unk word will use just unk score from kenlm (so kenlm has the unk word as normal word accumulating the full rest probability mass). In case of unk n-gram it is doing back-off on knowing (n-1)-gram + prior on the n-th word, and so on backing to 1-gram, where you have the full pdf. 

So you can train kenlm with the same lexicon but on larger corpus to capture more n-grams to have more precise probabilities (without doing backoff), however size of the kenlm also will be growing, so pay attention here because you can end up with >100GB bin file (like if you take common crawl and use only 200k dict for 4gram).

**Comment by abhinavkulkarni** [comment]
Thanks, @tlikhomanenko, I understand your comment about smearing using only unigram scores and `<unk>` token having a non-zero probability mass inside KenLM.

Let us say, I want to train an LM on Google News corpus which is quite huge. Let's say it has 100M unique words (post normalization), but I would only want to maintain a lexicon of 200k top occurring words due to runtime resource constraints.

In this case, do you first train a huge LM arpa on the whole corpus and then prune it so that:
1. any unigram that is **_NOT_** in the top 200k list is removed?
2. any n-gram that **_doesn't have all its words_** from the top 200k list is removed?

I took a glance at the source of [LexiconDecoder.cpp](https://github.com/facebookresearch/wav2letter/blob/v0.2/src/libraries/decoder/LexiconDecoder.cpp), and I can see this should work.

Let me know.

Thanks!

**Comment by trangtv57** [comment]
hi @abhinavkulkarni,
you can easy handle language model build by kenlm with limited vocab like script preprocess in lexiconfree-model.
[script to build limited vocab for kenlm](https://github.com/facebookresearch/wav2letter/blob/master/recipes/lexicon_free/librispeech/prepare_fairseq_data.sh)
and 
[script for building kenlm base limited vocab option in kenlm](https://github.com/facebookresearch/wav2letter/blob/master/recipes/lexicon_free/librispeech/train_ngram_lms.sh)
Im understand this is question you want to know, but not sure. If you have another discuss, just comment. tks

**Comment by tlikhomanenko** [comment]
@trangtv57 yep, this is correct comment, kenlm has an option to restrict vocab. This restriction works as 2. pointed by @abhinavkulkarni: so this is done on fly during computation that any ngram (even unigram) containing word not from the file provided will be not accounted. For my use-case I exactly used that. Kenlm don't have the case when you include all unigrams and you prun all others with restricted vocab.

Options:
- Train pruned arpa file with all words. Pruning here should be very aggressive
- Use SRILM (if I remember it has another logic on pruning and restriction in vocab compared to kenlm)
- do kenlm with restricted vocab but hack decoder to use unigrams anyway.
- do your own ngram (hack kenlm, SLIRM, ...)
